rust/operator/src/hdfs_controller.rs

```diff
@@ -0,0 +1,678 @@
+use lazy_static::lazy_static;
+use stackable_hdfs_crd::constants::*;
+use stackable_hdfs_crd::error::{Error, HdfsOperatorResult};
+use stackable_hdfs_crd::{HdfsCluster, HdfsPodRef, HdfsRole};
+use stackable_operator::builder::{ConfigMapBuilder, ObjectMetaBuilder};
+use stackable_operator::k8s_openapi::api::core::v1::{
+    Container, ContainerPort, HTTPGetAction, ObjectFieldSelector, PodSpec, PodTemplateSpec, Probe,
+    SecurityContext, VolumeMount,
+};
+use stackable_operator::k8s_openapi::api::{
+    apps::v1::{StatefulSet, StatefulSetSpec},
+    core::v1::{
+        ConfigMap, ConfigMapKeySelector, ConfigMapVolumeSource, EnvVar, EnvVarSource,
+        PersistentVolumeClaim, PersistentVolumeClaimSpec, ResourceRequirements, Service,
+        ServicePort, ServiceSpec, Volume,
+    },
+};
+use stackable_operator::k8s_openapi::apimachinery::pkg::util::intstr::IntOrString;
+use stackable_operator::k8s_openapi::apimachinery::pkg::{
+    api::resource::Quantity, apis::meta::v1::LabelSelector,
+};
+use stackable_operator::kube::runtime::reflector::ObjectRef;
+use stackable_operator::kube::{
+    api::ObjectMeta,
+    runtime::controller::{Context, ReconcilerAction},
+};
+use stackable_operator::labels::role_group_selector_labels;
+use stackable_operator::product_config::{types::PropertyNameKind, ProductConfigManager};
+use stackable_operator::product_config_utils::{
+    transform_all_roles_to_config, validate_all_roles_and_groups_config,
+};
+use stackable_operator::role_utils::RoleGroupRef;
+use std::collections::{BTreeMap, HashMap};
+use std::time::Duration;
+
+lazy_static! {
+    /// Liveliness probe used by all master, worker and history containers.
+    static ref PROBE: Probe = Probe {
+        http_get: Some(HTTPGetAction {
+            port: IntOrString::String(String::from(SERVICE_PORT_NAME_METRICS)),
+            ..HTTPGetAction::default()
+        }),
+        period_seconds: Some(10),
+        initial_delay_seconds: Some(10),
+        ..Probe::default()
+    };
+}
+
+pub struct Ctx {
+    pub client: stackable_operator::client::Client,
+    pub product_config: ProductConfigManager,
+}
+
+pub async fn reconcile_hdfs(
+    hdfs: HdfsCluster,
+    ctx: Context<Ctx>,
+) -> HdfsOperatorResult<ReconcilerAction> {
+    tracing::info!("Starting reconcile");
+    let client = &ctx.get_ref().client;
+
+    let validated_config = validate_all_roles_and_groups_config(
+        hdfs.hdfs_version()?,
+        &transform_all_roles_to_config(&hdfs, hdfs.build_role_properties()?)
+            .map_err(|source| Error::InvalidRoleConfig { source })?,
+        &ctx.get_ref().product_config,
+        false,
+        false,
+    )
+    .map_err(|source| Error::InvalidProductConfig { source })?;
```

Comment:
with `.with_context()` you can omit the source

Reply:
I don't use snafu here.